### ISSUE TITLE
Custom Endpoint Response Formatting

### ISSUE THREAD

**Issue by jhealey5** [WP-API/WP-API#2659]
I'm trying to use a custom endpoint (basically to get a random sorting working) and am using the following code:

```
    // Custom WP API endpoint
    function theme_enable_random_api() {

        // create json-api endpoint

        add_action('rest_api_init', function () {

            // http://example.com/wp-json/random/v2/posts

            register_rest_route('random/v2', '/random', array (
                'methods'             => 'GET',
                'callback'            => 'wp_json_offers_v2__posts',
                'permission_callback' => function (WP_REST_Request $request) {
                    return true;
                }
            ));
        });

        // handle the request

        function wp_json_offers_v2__posts($request) {
            // json-api params

            $parameters = $request->get_query_params();

            // default search args

            $args = array(
                'post_type'     => $parameters['type'],
                'numberposts'   => 9,
                'offset'        => $parameters['offset'],
                'post_not_in'       => $parameters['exclude'],
                'orderby'       => 'rand',
            );

            // run query

            $posts = get_posts($args);

            // return results
            return new WP_REST_Response($posts, 200);
        }

    }

    add_action('init', 'theme_enable_random_api');
```

However the response I'm getting is not the same as the response I would get from a standard call to the api.

Standard:
![image](https://cloud.githubusercontent.com/assets/1771431/17754563/08837c38-64cd-11e6-9fa2-ccba01498a44.png)

Custom Endpoint:
![image](https://cloud.githubusercontent.com/assets/1771431/17754579/14ab9dc4-64cd-11e6-865d-f19fb2fba3ba.png)

The problem really is that I can't access the taxonomy/acf information like I can in the standard one. I'm not great with PHP so more than likely I'm not getting it properly.

Cheers.

**Comment by websupporter** [comment]
Most likely, you are using a second plugin, which ensures the ACF meta data is send with the REST response. This plugin probably uses `register_rest_field` (see http://v2.wp-api.org/extending/modifying/) to add this data to the usual posts endpoint. You would need to add this data yourself. I guess the easiest way would be to somehow see, how the ACF2REST plugin is using `register_rest_field` and adding the data. From there you most likely get a good idea.

Basically, the idea of defining your own endpoint is: You are now in charge of defining what data will be passed back. If you want certain custom fields, you will need to add them. So: If you want to pass the attached terms, you will need to do so.

**Comment by jhealey5** [comment]
It seems the solution was in the form of:

```
 $data = $this->prepare_item_for_response( $post, $request );
 $posts[] = $this->prepare_response_for_collection( $data );
```

Which I guess preps the response and enables the hooks other plugins use to hook into it.

More on the stackoverflow answer: http://wordpress.stackexchange.com/questions/236249/wp-api-v2-custom-endpoint-response-formatting/236289#236289

Cheers!
